File: day11.py
import math

lst = []
#with open('testcase.txt', 'r') as file:
with open('11input.txt', 'r') as file:
    for line in file:
        ln = line.strip()
        lst = ln.split(" ")
        
col = {}

for val in lst:
    if (val in col.keys()):
        col[val] += 1
    else:
        col[val] = 1

print(col)

i = 0
n = 1000 #P1: 25, P2: 75
for i in range(n):
    ncol = {}
    for v in col.keys():
        if (v == "0"):
            if ("1" in ncol.keys()):
                ncol["1"] += col[v]
            else:
                ncol["1"] = col[v]
        elif (len(v)%2 == 0):
            y = v[0:math.floor(len(v)/2)]
            x = str(int(v[math.ceil(len(v)/2):]))
            if (y in ncol.keys()):
                ncol[y] += col[v]
            else:
                ncol[y] = col[v]
            if (x in ncol.keys()):
                ncol[x] += col[v]
            else:
                ncol[x] = col[v]
        else:
            x = str(int(v)*2024)
            if (x in ncol.keys()):
                ncol[x] += col[v]
            else:
                ncol[x] = col[v]
    col = ncol
       
num = 0
for v in col.keys():
    num += col[v]

print(f"Num stones: {num}")

File: day11.js
const fs = require('fs');
fs.readFile('11input.txt', 'utf8', (err, data) => {
//fs.readFile('testcase.txt', 'utf8', (err, data) => {
    lst = data.split(" ");

    col = {}
    for (i = 0; i < lst.length; i++) {
        val = lst[i];
        if (col[val]) {
            col[val] += 1
        } else {
            col[val] = 1
        }
    }

    n = 25
    for (i = 0; i < n; i++) {
        console.log(i)
        ncol = {}
        for (j = 0; j < Object.keys(col).length; j++) {
            v = Object.keys(col)[j]
            if (v == "0") {
                if (ncol["1"])
                    ncol["1"] += col[v]
                else
                    ncol["1"] = col[v]
            } else if (v.length%2 == 0) {
                y = v.substring(0, Math.floor(v.length/2))
                x = parseInt(v.substring(Math.ceil(v.length/2))).toString()
                if (ncol[y])
                    ncol[y] += col[v]
                else
                    ncol[y] = col[v]
                if (ncol[x])
                    ncol[x] += col[v]
                else
                    ncol[x] = col[v]
            } else {
                x = (parseInt(v)*2024).toString()
                if (ncol[x])
                    ncol[x] += col[v]
                else
                    ncol[x] = col[v]
            }
        }
        col = ncol
    }

    out = 0
    for (j = 0; j < Object.keys(col).length; j++) {
        out += col[Object.keys(col)[j]]
    }
    console.log(out)
});